algorithms/linfa-clustering/src/k_means/init.rs

```diff
@@ -0,0 +1,389 @@
+    // time. The values provided here are "sweetspots" suggested by the paper.
+    let n_rounds = 8;
+    let candidates_per_round = n_clusters;
+
+    let (n_samples, n_features) = observations.dim();
+    let mut candidates = Array2::zeros((n_clusters * n_rounds, n_features));
+
+    // Pick 1st centroid randomly
+    let first_idx = rng.gen_range(0, n_samples);
+    candidates.row_mut(0).assign(&observations.row(first_idx));
+    let mut n_candidates = 1;
+
+    let mut dists = Array1::zeros(n_samples);
+    'outer: for _ in 0..n_rounds {
+        let current_candidates = candidates.slice(s![0..n_candidates, ..]);
+        update_min_dists(&current_candidates, &observations, &mut dists);
+        // Generate the next set of candidates from the input points, using the same probability
+        // formula as KMeans++. On average this generates candidates equal to
+        // `candidates_per_round`.
+        let next_candidates = sample_subsequent_candidates(
+            observations,
+            &dists,
+            F::from(candidates_per_round).unwrap(),
+            rng.gen_range(0, 100000),
+            rng_func,
+        );
+
+        // Append the newly generated candidates to the current cadidates, breaking out of the loop
+        // if too many candidates have been found
+        for candidate in next_candidates.into_iter() {
+            candidates.row_mut(n_candidates).assign(&candidate);
+            n_candidates += 1;
+            if n_candidates >= candidates.nrows() {
+                break 'outer;
+            }
+        }
+    }
+
+    let final_candidates = candidates.slice(s![0..n_candidates, ..]);
+    // Weigh the candidate centroids by the sizes of the clusters they form in the input points.
+    let weights = cluster_membership_counts(&final_candidates, &observations);
+
+    // The number of candidates is almost certainly higher than the number of centroids, so we
+    // recluster the candidates into the right number of centroids using weighted KMeans++.
+    weighted_k_means_plusplus(n_clusters, &final_candidates, &weights.view(), rng)
+}
+
+/// Generate candidate centroids by sampling each observation in parallel using a supplied random
+/// number generation function.
+fn sample_subsequent_candidates<'a, F: Float>(
+    observations: &'a ArrayView2<'a, F>,
+    dists: &Array1<F>,
+    multiplier: F,
+    seed: u64,
+    // Must return a number between 0 and 1.
+    // Easiest way to generate a random number in a Rayon thread, since cloning the actual RNG is
+    // expensive.
+    rng_func: &RngFunc<F>,
+) -> Vec<ArrayView1<'a, F>> {
+    // This sum can also be parallelized
+    let cost = dists.sum();
+
+    // The sequential alternative to this operation is to taking "multiplier" samples from a
+    // weighted index of "dists". Doing so avoids using rng_func but may lead to slower code.
+    dists
+        .axis_iter(Axis(0))
```

Comment:
I need `.axis_iter()` to produce an indexed parallel iterator. If I used `iter()` then the parallel iterator wouldn't be index and I wouldn't be able to call `enumerate()` on it.

Reply:
I don't understand, `axis_iter` is not creating a parallel iterator and according to the [docs](https://docs.rs/ndarray/0.13.0/ndarray/parallel/index.html#arrays-and-array-views) you can call `par_iter` to iterate over the elements of an array

Reply:
[Here](https://docs.rs/ndarray/0.13.0/ndarray/parallel/index.html) it mentions that you have to call `into_par_iter()` on the results of `axis_iter()` to produce an indexed parallel iterator. If you call `par_iter()` on non-axis iterators then you get a non-indexed parallel iterator.

Reply:
damn rayon's iterators are more complicated than thought, but it makes sense to me know. You need the indexed parallel trait bound for the enumeration of items, otherwise the length [here](https://github.com/rayon-rs/rayon/blob/1f069d77101bbc42e4a42c4a5aec16b326dc4e32/src/iter/enumerate.rs#L109) is not known prior. You can do
```rust
dists.as_slice().unwrap()
  .into_par_iter()
  .enumerate()
  .filter_map(move |(i, d)| {
    ....
  })
```
but this is only a bit more concise